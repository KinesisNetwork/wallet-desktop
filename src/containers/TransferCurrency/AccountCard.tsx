import { AddressView } from '@components/AddressView'
import { InitialsAvatar } from '@components/InitialsAvatar'
import { AddressDisplay } from '@types'
import * as React from 'react'

interface TransferFromAccount {
  name: string
  address: string
  addressDisplay: AddressDisplay
}

export const AccountCard: React.SFC<TransferFromAccount> = (props: TransferFromAccount) => {
  return (
    <div className="column">
      <div className="box">
        <div className="level">
          <div className="level-left">
            <div className="level-item">
              <InitialsAvatar name={props.name} size="small" />
            </div>
            <div className="level-item">
              <p className="has-text-grey-lighter">
                <AddressView address={props.address} addressDisplay={props.addressDisplay} />
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}
