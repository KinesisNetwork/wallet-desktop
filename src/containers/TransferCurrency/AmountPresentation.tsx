import { addMetalColour } from '@helpers/walletUtils'
import * as React from 'react'

interface Amount {
  amount: string | number
  text: string
  currency: string
  isTestnet?: boolean
}

export const AmountPresentation: React.SFC<Amount> = (props: Amount) => {
  const decimalPlaces = props.currency === 'KEM' ? 7 : 5
  return (
    <div className="level">
      <div className="level-item">
        <h1 className={`title is-size-2 has-text-weight-bold ${addMetalColour(props.currency)}`}>
          {Number(props.amount).toFixed(decimalPlaces)}{' '}
          {props.isTestnet ? 'T' + props.text : props.text}
        </h1>
      </div>
    </div>
  )
}
