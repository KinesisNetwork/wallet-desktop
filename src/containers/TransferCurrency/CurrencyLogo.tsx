import * as React from 'react'
import SVG from 'react-inlinesvg'

import { setImageSize } from '@services/util'
import { Currency, ImageSize } from '@types'
import * as kagLogo from 'images/kag-icon.svg'
import * as kauLogo from 'images/kau-icon.svg'

interface Props {
  currency: string
  size: ImageSize
  title?: string
}

export const CurrencyLogo: React.SFC<Props> = props => {
  return (
    <React.Fragment>
      <div className="level">
        <div className="level-item">
          <figure className={`image ${setImageSize(props.size).image}`}>
            <SVG src={props.currency === Currency.KAU ? kauLogo : kagLogo} />
          </figure>
        </div>
      </div>
      {props.title && (
        <div className="level" style={{ letterSpacing: '2px' }}>
          <div className="level-item">
            <h1 className="title has-text-weight-bold is-uppercase">{props.title}</h1>
          </div>
        </div>
      )}
    </React.Fragment>
  )
}
