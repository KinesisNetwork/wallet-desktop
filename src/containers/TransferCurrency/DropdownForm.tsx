import * as React from 'react'

import { DropdownDivider } from '@containers/Sidebar/DropdownDivider'
import { getInactiveAccountsInContactFormat } from '@services/accounts'
import { Contact, FormChangeHandler, TransferRequest, WalletAccount } from '@types'

interface Props {
  onFieldChange: () => void
  savedContacts: Contact[]
  payeePublicKey: TransferRequest['targetPayee']
  handleChange: FormChangeHandler<TransferRequest>
  accounts: WalletAccount[]
  activeAccount: WalletAccount
}

export const DropdownFormPresentation: React.SFC<Props> = props => {
  const inactiveAccounts = getInactiveAccountsInContactFormat(props.accounts, props.activeAccount)

  const renderOption = (contactsToOption: Contact[]) => {
    return contactsToOption.map(({ name, address }) => (
      <option key={address} value={address}>
        {name}
      </option>
    ))
  }

  const isOnContactList = props.savedContacts
    .concat(inactiveAccounts)
    .find(({ address }) => props.payeePublicKey === address)

  return (
    <React.Fragment>
      <label className="label is-small">To</label>
      <div className="field is-grouped">
        <div className="control is-expanded">
          <div className="select is-fullwidth payee-select">
            <select
              className={!isOnContactList ? 'has-text-grey' : ''}
              onChange={({ target: { value: newValue } }) =>
                props.handleChange({ field: 'targetPayee', newValue })
              }
              value={props.payeePublicKey}
            >
              <option>Select a contact</option>
              {renderOption(props.savedContacts)}
              {props.accounts.length > 1 && (
                <React.Fragment>
                  <option disabled={true}>
                    <DropdownDivider />
                  </option>
                  {renderOption(inactiveAccounts)}
                </React.Fragment>
              )}
            </select>
          </div>
        </div>
        <div className="control is-flex">
          <span
            className="is-uppercase is-size-7 has-text-grey-lighter"
            style={{ alignSelf: 'center' }}
          >
            or
          </span>
        </div>
        <div className="control">
          <button className="button is-primary" onClick={props.onFieldChange}>
            Add new
          </button>
        </div>
      </div>
    </React.Fragment>
  )
}

export { DropdownFormPresentation as DropdownForm }
