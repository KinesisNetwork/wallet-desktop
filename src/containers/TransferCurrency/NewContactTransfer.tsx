import { InputField } from '@components/InputField'
import { Contact, FormErrors } from '@types'
import * as React from 'react'

interface Props {
  onFieldChange: () => void
  handleChange: (field: keyof Contact, value: string) => void
  errors: FormErrors
  onSaveToContactsChange: () => void
  saveToContacts: boolean
  name: string
  publicKey: string
}

export class NewContactTransferPresentation extends React.Component<Props> {
  render() {
    return (
      <React.Fragment>
        <div className="field" style={{ position: 'relative' }}>
          <button
            className="button is-outlined is-light"
            style={{ position: 'absolute', left: '-15%', top: '1.47rem' }}
            onClick={this.props.onFieldChange}
          >
            <span className="icon">
              <i className="fal fa-lg fa-angle-left" />
            </span>
          </button>
          <InputField
            label="Contact name"
            value={this.props.name}
            onChangeHandler={value => this.props.handleChange('name', value)}
            id="contact-name"
            icon="fa-user-circle"
            name="name"
            disabled={!this.props.saveToContacts}
          />
          <InputField
            label="Kinesis Address"
            value={this.props.publicKey}
            placeholder="example: &quot;GDUL65KWQ4PJA7FLO6467...CM7G&quot;"
            onChangeHandler={value => this.props.handleChange('address', value)}
            id="kinesis-address"
            icon="fa-qrcode"
            errorText={this.props.errors.targetPayee}
          />
          <div className="level">
            <div className="level-left">
              <div className="field">
                <input
                  id="switchRoundedSuccess"
                  type="checkbox"
                  name="switchRoundedSuccess"
                  className="switch is-rounded is-success"
                  onChange={this.props.onSaveToContactsChange}
                  checked={this.props.saveToContacts}
                />
                <label htmlFor="switchRoundedSuccess">Save to contacts</label>
              </div>
            </div>
          </div>
        </div>
      </React.Fragment>
    )
  }
}

export { NewContactTransferPresentation as NewContactTransfer }
