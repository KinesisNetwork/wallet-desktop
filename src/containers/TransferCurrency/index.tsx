export { TransferPage } from './TransferPage'
export { ConfirmationPage } from './ConfirmationPage'
