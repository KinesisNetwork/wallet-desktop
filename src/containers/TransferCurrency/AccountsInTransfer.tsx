import { AccountCard } from '@containers/TransferCurrency/AccountCard'
import { getActiveAccount } from '@selectors'
import { getInactiveAccounts } from '@services/accounts'
import { RootState } from '@store'
import { AddressDisplay } from '@types'
import * as React from 'react'
import { connect } from 'react-redux'

const mapStateToProps = ({ wallet, transfer, contacts }: RootState) => ({
  formData: transfer.formData,
  walletName: wallet.persisted.walletName,
  contactList: contacts.contactList,
  activeAccount: getActiveAccount(wallet),
  accounts: wallet.accounts,
})

type Props = ReturnType<typeof mapStateToProps>

export const AccountsInTransferPresentation: React.SFC<Props> = props => {
  const inactiveAccounts = getInactiveAccounts(props.accounts, props.activeAccount)

  const getPayeeNameForAvatar = () => {
    const payeeAddress = props.formData.targetPayee
    const payee = props.contactList
      .concat(inactiveAccounts)
      .find(contact => contact.address === payeeAddress)
    return payee ? payee.name : payeeAddress
  }

  return (
    <React.Fragment>
      <AccountCard
        name={props.walletName}
        addressDisplay={AddressDisplay.account}
        address={props.activeAccount.keypair.publicKey()}
      />
      <div className="column is-narrow">
        <div className="level">
          <div className="level-item">
            <span className="has-text-grey-lighter is-size-2">
              <i className="fal fa-arrow-circle-right" />
            </span>
          </div>
        </div>
      </div>
      <AccountCard
        name={getPayeeNameForAvatar()}
        address={props.formData.targetPayee}
        addressDisplay={AddressDisplay.payee}
      />
    </React.Fragment>
  )
}

const AccountsInTransfer = connect(mapStateToProps)(AccountsInTransferPresentation)

export { AccountsInTransfer }
