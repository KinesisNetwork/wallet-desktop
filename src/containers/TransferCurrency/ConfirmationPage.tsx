import { RootState } from '@store'
import * as React from 'react'
import { connect } from 'react-redux'

import * as kagLogo from '@icons/kag-icon.svg'
import * as kauLogo from '@icons/kau-icon.svg'

import { showNotification, transferRequest } from '@actions'
import { AccountCard } from '@containers/TransferCurrency/AccountCard'
import { TransferSummary } from '@containers/TransferCurrency/TransferSummary'
import { addMetalColour } from '@helpers/walletUtils'
import { AddressDisplay, Currency, RootRoutes } from '@types'
import { goBack, replace } from 'connected-react-router'

import { Loader } from '@components/Loader'
import { getActiveAccount } from '@selectors'

const mapStateToProps = (state: RootState) => {
  const {
    transfer: { formData },
    wallet,
  } = state
  return {
    currency: state.connections.currentCurrency,
    walletName: state.wallet.persisted.walletName,
    memo: formData.memo,
    fee: formData.fee,
    amount: formData.amount,
    formData,
    activeAccount: getActiveAccount(wallet),
    isTransferring: state.transfer.isTransferring,
    contactList: state.contacts.contactList,
  }
}

const mapDispatchToProps = {
  goBackToTransformPage: () => goBack(),
  transferRequest,
  goToDashboard: () => replace(RootRoutes.dashboard),
  showNotification,
}

type Props = ReturnType<typeof mapStateToProps> & typeof mapDispatchToProps

export class ConfirmationPagePresentation extends React.Component<Props> {
  confirmAndGoToDashboard = () => {
    this.props.transferRequest(this.props.formData)
  }

  getPayeeNameForAvatar = () => {
    const payeeAddress = this.props.formData.targetPayee
    const payee = this.props.contactList.find(contact => contact.address === payeeAddress)
    return payee ? payee.name : payeeAddress
  }

  render() {
    return (
      <div style={{ position: 'relative' }}>
        {this.props.isTransferring && <Loader />}
        <div
          className="columns is-mobile is-centered"
          style={this.props.isTransferring ? { filter: 'blur(2px)' } : {}}
        >
          <div className="column is-three-fifths">
            <section className="section has-centered">
              <div className="level">
                <div className="level-item">
                  <figure className="image is-64x64">
                    <img
                      src={this.props.currency === Currency.KAU ? kauLogo : kagLogo}
                      className="is-rounded"
                    />
                  </figure>
                </div>
              </div>
              <div className="level">
                <div className="level-item">
                  <h1 className="title has-text-weight-bold is-uppercase">Confirm transfer</h1>
                </div>
              </div>
            </section>
            <section className="columns is-vcentered">
              <AccountCard
                name={this.props.walletName}
                addressDisplay={AddressDisplay.account}
                address={this.props.activeAccount.keypair.publicKey()}
              />
              <div className="column is-narrow">
                <div className="level">
                  <div className="level-item">
                    <span className="has-text-grey-lighter is-size-2">
                      <i className="fal fa-arrow-circle-right" />
                    </span>
                  </div>
                </div>
              </div>
              <AccountCard
                name={this.getPayeeNameForAvatar()}
                address={this.props.formData.targetPayee}
                addressDisplay={AddressDisplay.payee}
              />
            </section>
            <section className="section">
              <div className="columns is-centered">
                <div className="column is-two-thirds">
                  <div className="has-text-centered content">
                    <h1
                      className={`is-size-1 has-text-weight-bold ${addMetalColour(
                        this.props.currency,
                      )}`}
                    >
                      {Number(this.props.amount).toFixed(5)} {this.props.currency}
                    </h1>
                    <p className="has-text-grey-lighter">
                      {this.props.memo && <q>{this.props.memo}</q>}
                    </p>
                  </div>
                  <hr className="has-background-grey-lighter" />
                  <TransferSummary
                    currency={this.props.currency}
                    description="Transaction fee"
                    amount={Number(this.props.fee)}
                  />
                  <hr className="has-background-grey-lighter" />
                  <TransferSummary
                    currency={this.props.currency}
                    description="TOTAL"
                    amount={Number(this.props.amount) + Number(this.props.fee)}
                  />
                  <hr className="has-background-grey-lighter" />
                  <section className="field is-grouped is-grouped-right">
                    <p className="control">
                      <button className="button is-text" onClick={this.props.goBackToTransformPage}>
                        Back
                      </button>
                    </p>
                    <p className="control">
                      <button className="button is-primary" onClick={this.confirmAndGoToDashboard}>
                        <span className="icon">
                          <i className="fal fa-arrow-up" />
                        </span>
                        <span>Confirm</span>
                      </button>
                    </p>
                  </section>
                </div>
              </div>
            </section>
          </div>
        </div>
      </div>
    )
  }
}

const ConnectedConfirmationPage = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConfirmationPagePresentation)

export { ConnectedConfirmationPage as ConfirmationPage }
