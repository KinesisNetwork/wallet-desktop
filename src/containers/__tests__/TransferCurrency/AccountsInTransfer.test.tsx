import { AccountsInTransferPresentation } from '@containers/TransferCurrency/AccountsInTransfer'
import { shallow } from 'enzyme'
import * as React from 'react'
import '../../../setupTests'

describe('AccountsInTransfer', () => {
  let props

  beforeEach(() => {
    props = {
      formData: {
        targetPayee: '123qwe',
        amount: '0.5',
        memo: 'transfer',
        fee: '0.0002',
      },
      wallerName: 'my wallet',
      contactList: [
        {
          name: 'John',
          address: '123qwe',
        },
      ],
      activeAccount: {
        name: 'Account 1',
        keypair: {
          publicKey: () => null,
        },
      },
    }
  })

  it('should render correctly', () => {
    const wrapper = shallow(<AccountsInTransferPresentation {...props} />)

    expect(wrapper).toMatchSnapshot()
  })
})
