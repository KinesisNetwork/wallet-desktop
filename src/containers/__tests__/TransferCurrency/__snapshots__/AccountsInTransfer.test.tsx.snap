// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsInTransfer should render correctly 1`] = `
<Fragment>
  <Component
    address={null}
    addressDisplay={0}
  />
  <div
    className="column is-narrow"
  >
    <div
      className="level"
    >
      <div
        className="level-item"
      >
        <span
          className="has-text-grey-lighter is-size-2"
        >
          <i
            className="fal fa-arrow-circle-right"
          />
        </span>
      </div>
    </div>
  </div>
  <Component
    address="123qwe"
    addressDisplay={1}
    name="John"
  />
</Fragment>
`;
